sharktank/sharktank/layers/ffn_moe_block.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2024 Advanced Micro Devices, Inc
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from typing import Optional
+
+import torch
+import torch.nn.functional as F
+
+from .base import ThetaLayer
+from .linear import LinearLayer
+from ..types import Theta, DefaultPrimitiveTensor
+
+__all__ = [
+    "FFNMOE",
+]
+
+
+class FFNMOE(ThetaLayer):
+    def __init__(
+        self,
+        theta: Theta,
+        expert_idx: Optional[int] = None,
+    ):
+
+        super().__init__(theta)
+
+        if theta.optional_tensor("ffn_gate_exps") is not None:
+            merged_tensor = theta.tensor("ffn_gate_exps", "weight")
+
+            expert_tensor = extract_ffn_layer(
+                merged_tensor=merged_tensor,
+                layer_name="ffn_gate",
+                expert_idx=expert_idx,
+            )
+
+            self.add_module("ffn_gate", LinearLayer(Theta({"weight": expert_tensor})))
+
+            merged_tensor = theta.tensor("ffn_up_exps", "weight")
+
+            expert_tensor = extract_ffn_layer(
+                merged_tensor=merged_tensor, layer_name="ffn_up", expert_idx=expert_idx
+            )
+
+            self.add_module("ffn_up", LinearLayer(Theta({"weight": expert_tensor})))
+
+            merged_tensor = theta.tensor("ffn_down_exps", "weight")
+
+            expert_tensor = extract_ffn_layer(
+                merged_tensor=merged_tensor,
+                layer_name="ffn_down",
+                expert_idx=expert_idx,
+            )
+
+            self.add_module("ffn_down", LinearLayer(Theta({"weight": expert_tensor})))
+
+        else:
+            self.add_module("ffn_gate", LinearLayer(theta("ffn_gate", expert_idx)))
+            self.add_module("ffn_up", LinearLayer(theta("ffn_up", expert_idx)))
+            self.add_module("ffn_down", LinearLayer(theta("ffn_down", expert_idx)))
+
+    def forward(
+        self,
+        h: torch.Tensor,
+    ):
+        ffn_gate = F.silu(self.ffn_gate(h))
+        ffn_up = self.ffn_up(h)
+        ffn_down = self.ffn_down(ffn_gate * ffn_up)
+        return ffn_down
+
+
+def extract_ffn_layer(
```

Comment:
Please adopt this function from [here](https://github.com/archana-ramalingam/sharktank/blob/mixtral/sharktank/sharktank/layers/ffn_moe_block.py#L74). This is necessary to split expert tensors by their names for MOE gguf models.

Reply:
I don't think the name of the primitiveTensor matters, do you?

Reply:
It matters when debugging/tracing theta, to know which ffn layer/expert and block this belongs to, as this is fed to theta in the FFNMOE class above. Like I said we are creating new layer names for each of 8 ffn expert layers from merged ffn expert layers.